src/time.c

```diff
@@ -0,0 +1,276 @@
+
+int
+tm_to_efi_time(const struct tm * const s, efi_time_t *d, bool tzadj)
+{
+	if (!s || !d) {
+		errno = EINVAL;
+		return -1;
+	}
+
+	d->pad2 = 0;
+	d->daylight = s->tm_isdst ? EFI_TIME_IN_DAYLIGHT : 0;
+	d->timezone = 0;
+	d->nanosecond = 0;
+	d->pad1 = 0;
+	/*
+	 * Just ignore EFI's range problem here and pretend we're in UTC
+	 * not UT1.
+	 */
+	d->second = s->tm_sec;
+	d->minute = s->tm_min;
+	d->hour = s->tm_hour;
+	d->day = s->tm_mday;
+	d->month = s->tm_mon + 1;
+	d->year = s->tm_year + 1900;
+
+	if (tzadj) {
+		tzset();
+		d->timezone = timezone / 60;
+	}
+
+	return 0;
+}
+
+static char *otz_;
+static char *ntz_;
+
+static const char *
+newtz(int16_t timezone_)
+{
+	if (!otz_)
+		otz_ = strdup(secure_getenv("TZ"));
+
+	if (ntz_) {
+		free(ntz_);
+		ntz_ = NULL;
+	}
+
+	if (timezone_ == EFI_UNSPECIFIED_TIMEZONE) {
+		unsetenv("TZ");
+	} else {
+		char tzsign = timezone_ >= 0 ? '+' : '-';
+		int tzabs = tzsign == '+' ? timezone_ : -timezone_;
+		int16_t tzhours = tzabs / 60;
+		int16_t tzminutes = tzabs % 60;
+		int rc;
+
+		/*
+		 * I have no idea what the right thing to do with DST is
+		 * here, so I'm going to ignore it.
+		 */
+		rc = asprintf(&ntz_, "UTC%c%"PRId16":%"PRId16":00",
+			  tzsign, tzhours, tzminutes);
+		if (rc < 1)
+			return NULL;
+
+		setenv("TZ", ntz_, 1);
+	}
+	tzset();
+
+	return ntz_;
+}
+
+static const char *
+oldtz(void) {
+	if (ntz_) {
+		free(ntz_);
+		ntz_ = NULL;
+
+		if (otz_)
+			setenv("TZ", otz_, 1);
+		else
+			unsetenv("TZ");
+	}
+
+	tzset();
+
+	return otz_;
+}
+
+efi_time_t *
+efi_gmtime_r(const time_t *time, efi_time_t *result)
+{
+	struct tm tm = { 0 };
+
+	if (!time || !result) {
+		errno = EINVAL;
+		return NULL;
+	}
+
+	gmtime_r(time, &tm);
+	tm_to_efi_time(&tm, result, false);
+
+	return result;
+}
+
+efi_time_t *
+efi_gmtime(const time_t *time)
+{
+	static efi_time_t ret;
+
+	if (!time) {
+		errno = EINVAL;
+		return NULL;
+	}
+
+	efi_gmtime_r(time, &ret);
+
+	return &ret;
+}
+
+efi_time_t *
+efi_localtime_r(const time_t *time, efi_time_t *result)
+{
+	struct tm tm = { 0 };
+
+	if (!time || !result) {
+		errno = EINVAL;
+		return NULL;
+	}
+
+	localtime_r(time, &tm);
+	tm_to_efi_time(&tm, result, true);
+
+	return result;
+}
+
+efi_time_t *
+efi_localtime(const time_t *time)
+{
+	static efi_time_t ret;
+
+	if (!time) {
+		errno = EINVAL;
+		return NULL;
+	}
+
+	efi_localtime_r(time, &ret);
+
+	return &ret;
+}
+
+time_t
+efi_mktime(const efi_time_t * const time)
+{
+	struct tm tm = { 0 };
+	time_t ret;
+
+	if (!time) {
+		errno = EINVAL;
+		return (time_t)-1;
+	}
+
+	newtz(time->timezone);
+
+	efi_time_to_tm(time, &tm);
+	ret = mktime(&tm);
+
+	oldtz();
+
+	return ret;
+}
+
+char *
+efi_strptime(const char *s, const char *format, efi_time_t *time)
+{
+	struct tm tm;
+	char *end;
+
+	if (!s || !format || !time) {
+		errno = EINVAL;
+		return NULL;
+	}
+
+	memset(&tm, 0, sizeof(tm));
+	end = strptime(s, format, &tm);
+	if (end != NULL && tm_to_efi_time(&tm, time, true) < 0)
+		return NULL;
```

Comment:
Should this be `end` instead of `NULL` here?  A `NULL` return from regular strptime() is an error state.  I would have expected something like:

```C
end = strptime(s, format, &tm);
if (end == NULL || tm_to_efi_time(&tm, time, true) < 0)
    return NULL;
return end;
```

Reply:
It seems correct to me?  The `end != NULL` check is to only call `tm_to_efi_time()` if `strptime()` succeeded, and the `return NULL;` is because `end` isn't valid if `tm_to_efi_time()` failed (which in this case means `time` was `NULL`, so we won't actually get here because that was checked above.)  Would it be clearer if it were like this?
```
end = strptime(s, format, &tm);
if (end != NULL) {
    int rc = tm_to_efi_time(&tm, time, true);
    if (rc < 0)
        return NULL;
}
return end;
```
or maybe this?
```
end = strptime(s, format, &tm);
if (end == NULL)
    return NULL;

if (tm_to_efi_time(&tm, time, true) < 0)
    return NULL;

return end;
```

Reply:
I've updated it to use the last version above.

Reply:
Having slept on it, I think you're right, but I do prefer the current version to what was originally there.